File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>res;
    void help(int index , int target , vector<int>&nums , vector<int>&path){
        if(index == nums.size()){
            if(target==0){
                res.push_back(path);
                return;
            }
            else{
                return;
            }
        }
        
        if(target-nums[index]>=0){
            path.push_back(nums[index]);
            help(index,target-nums[index],nums,path);
            path.pop_back();
        }
        
        help(index+1,target-0,nums,path);
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int>path;
        help(0,target,nums,path);
        return res;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    
    
    List<List<Integer>>res = new ArrayList<>();
    
    public void help(int index , int[] nums , ArrayList<Integer> path , int target){
        
        if(index == nums.length && target == 0){
            res.add(new ArrayList(path));
            return;
        }
        
        if(index >= nums.length || target < 0){
            return;
        }
        
        help(index+1,nums,path,target-0);
        
        if(target - nums[index] >= 0){
            path.add(nums[index]);
            help(index,nums,path,target-nums[index]);
            path.remove(path.size()-1);
        }
        
    }
    
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        int index = 0;
        ArrayList<Integer> path = new ArrayList<>();
        help(index,nums,path,target);
        return res;
    }
}